src/main/java/io/gravitee/llama/cpp/LlamaIterator.java

```diff
@@ -16,165 +16,293 @@
+   * @return true if the state should continue, false if it should finish
+   */
+  protected boolean shouldContinue(ConversationState state, int tokenId) {
+    var tokenizer = state.getTokenizer();
 
-  public GenerationState getState() {
-    return state;
-  }
+    // Check for end-of-generation token
+    if (tokenizer.isEog(tokenId)) {
+      state.setFinishReason(FinishReason.STOP);
+      return false;
+    }
 
-  public LlamaIterator setStopStrings(List<String> stopStrings) {
-    stopString.initialize(stopStrings);
-    int maxStringSize = stopStrings.stream().mapToInt(String::length).max().orElse(0);
-    promptMemory.initialize(maxStringSize);
-    return this;
+    // Check token limit
+    int maxTokens = state.getMaxTokens();
+    if (maxTokens != -1 && maxTokens <= state.getAnswerTokens()) {
+      state.setFinishReason(FinishReason.LENGTH);
+      return false;
+    }
+
+    return true;
   }
 
-  public LlamaIterator setReasoning(String tokenStart, String tokenEnd) {
-    this.states.add(new StateBounds(REASONING, tokenStart, tokenEnd));
-    return this;
+  /**
+   * Helper methods for finish reason detection.
+   */
+  protected boolean isEog(int tokenId) {
+    boolean isEog = currentState.getTokenizer().isEog(tokenId);
+    if (isEog) {
+      setFinishReason(STOP);
+    }
+    return isEog;
   }
 
-  public LlamaIterator setToolCall(String tokenStart, String tokenEnd) {
-    this.states.add(new StateBounds(TOOLS, tokenStart, tokenEnd));
-    return this;
+  protected boolean hasNotReachedQuota() {
+    int maxTokens = currentState.getMaxTokens();
+    boolean hasNotReachedQuota = maxTokens == -1 || maxTokens > currentState.getAnswerTokens();
+    if (!hasNotReachedQuota) {
+      setFinishReason(LENGTH);
+    }
+    return hasNotReachedQuota;
   }
 
   protected boolean endWithStopString() {
-    if (!promptMemory.isInitialized()) {
+    if (!currentState.getPromptMemory().isInitialized()) {
       return false;
     }
 
-    boolean endsWithStopString = stopString.evaluate(promptMemory.getMemory());
+    boolean endsWithStopString = currentState.getStopString().evaluate(currentState.getPromptMemory().getMemory());
     if (endsWithStopString) {
       setFinishReason(STOP);
     }
     return endsWithStopString;
   }
 
   protected void setFinishReason(FinishReason finishReason) {
-    if (this.finishReason != null) {
-      // We set this.finishReason here only if it has not been set to TOOL_CALL
-      // We make an exception only if the current finishReason to set is LENGTH
-      if (!TOOL_CALL.equals(this.finishReason) || LENGTH.equals(finishReason)) {
-        this.finishReason = finishReason;
+    if (currentState.getFinishReason() != null) {
+      if (!TOOL_CALL.equals(currentState.getFinishReason()) || LENGTH.equals(finishReason)) {
+        currentState.setFinishReason(finishReason);
       }
     } else {
-      this.finishReason = finishReason;
+      currentState.setFinishReason(finishReason);
     }
```

Comment:
Is the logic here essentially...

if (currentState.getFinishReason() == null || !TOOL_CALL.equals(currentState.getFinishReason()) || LENGTH.equals(finishReason)) { curentState.setFinishReason(finishReason) }

I think that captures the same thing... I'm not at all sure it's any easier to parse, I just got caught up for a second tracking back if the else clause was applying to the first or second if 🤣